security-archive: add a security db & cache system

ruyadorno opened the pull request:
Add a new package `@vltpkg/security-archive` that handles accessing, fetching remote data and persisting security report data on packages from a given graph.

## Usage

```js
import { actual } from '@vltpkg/graph'
import { SecurityArchive } from '@vltpkg/security-archive'
const specOptions = {
  registry: 'https://registry.npmjs.org/',
}
const graph = actual.load({
  ...specOptions,
  projectRoot: process.cwd(),
})
const archive = await SecurityArchive.start({ graph, specOptions })
if (archive.ok) {
  for (const node of graph.nodes.values()) {
    const securityData = archive.get(node.id)
    if (securityData) {
      console.log('securityData', securityData)
    }
  }
} else {
  console.warn('Failed to start the SecurityArchive')
}
```

ruyadorno commented on src/security-archive/src/index.ts:
```diff
@@ -0,0 +1,291 @@
+}
+
+// Loads the version number to be used in the User-Agent header
+const { version } = loadPackageJson(import.meta.filename) as {
+  version: string
+}
```
Need to use the special env var here for builds, e.g:
```suggestion
export const { version } = loadPackageJson(
  import.meta.filename,
  process.env.__VLT_INTERNAL_CLI_PACKAGE_JSON,
) as {
  version: string
}
```

isaacs commented on src/security-archive/src/index.ts:
```diff
@@ -0,0 +1,297 @@
+    for (const [key, value] of entries) {
+      insertData += `('${key}', '${JSON.stringify(value)}'),`
+    }
+    insertData = insertData.substring(0, insertData.length - 1)
```
This feels a bit clunky with the repeated appends and then substring at the end?

Consider:

```ts
const insertData: string[] = []
for (const [key, value] of entries) {
  insertData.push(`('${key}', '${JSON.stringify(value)}')`)
}
const dbWrite = db.prepare(
  `INSERT OR REPLACE INTO cache (key, value) VALUES ${insertData.join(',')}`,
)
dbWrite.run()
```

ruyadorno replied:
of course!

isaacs commented on src/security-archive/src/index.ts:
```diff
@@ -0,0 +1,297 @@
+  value: PackageReportData
+}
+
+export type SecurityArchiveOptions = {
```
This should extend `LRUOptions`, so that we can set the max, ttl, etc.

If anything really should not be set, then you can make it only settable to `undefined` in this type, like `fetchMethod?: undefined`.

ruyadorno replied:
I'll go ahead and add that suggestion to the implementation to be explicit about not supporting `fetchMethod`.

isaacs commented on src/security-archive/src/index.ts:
```diff
@@ -0,0 +1,297 @@
+  constructor(options: SecurityArchiveOptions = {}) {
+    super({
+      max: SecurityArchive.defaultMax,
+      ttl: SecurityArchive.defaultTtl,
```
```suggestion
      ttl: SecurityArchive.defaultTtl,
      ...options,
```

isaacs commented on src/security-archive/src/index.ts:
```diff
@@ -0,0 +1,297 @@
+ * using the SecurityArchive.get() method.
+ */
+export class SecurityArchive
+  extends LRUCache<DepID, PackageReportData>
```
If I'm reading this right, it seems like it loads _every_ security notice we've ever seen, every time it's used?

It seems like you could get better overall performance by leveraging the `fetchMethod` LRUCache behavior. That could make this more efficient in some cases, by avoiding loading data we don't end up needing.

ruyadorno replied:
currently it loads every security notice for the set of nodes found in the `graph` passed in the `refresh` method.

ruyadorno replied:
I don't think `fetchMethod` is going to be the best tool in the box for this particular task at hand. We do want to leverage the batch API Request, reading all packages all at once fits better that same model.

One more limitation / design goal that I'm going for in this current implementation that might not be straightforward is the fact that the SocketAPI can only be called from the backend (mostly due to a lack of CORS header in the server response) - that makes it so that the isomorphic implementation (at `./src/browser.ts`) can't really rely on a persistency + fetch API model that is available on node/backend.

The solution on my design is having the isomorphic implementation be a [standard map](https://github.com/vltpkg/vltpkg/pull/518/files#diff-df31307ba612f9fe6af94c88acf8ab099e483c379ac09cb781bc017be5763524R34-R37) that gets populated with data as part of the initial [`TransferData`](https://github.com/vltpkg/vltpkg/blob/main/src/gui/src/state/types.ts#L55-L59) (pratically speaking, this is the `graph.json` returned by the GUI server). It will make a snappier end-user experience in the frontend (which matters a lot since users are expected to be tweaking the search bar a lot) while having a tradeoff of having to load upfront info on all packages for that given graph, which again, in the frontend the likeness that users are going to tweak the searchbar to `*`, `:root > *` is very high, so having that information upfront makes sense.

isaacs commented on src/security-archive/src/index.ts:
```diff
@@ -0,0 +1,297 @@
+        start,
+      })
+    }
+    this.purgeStale()
```
Why is this needed? Can we avoid it by just not adding anything that we know is past its TTL?

ruyadorno replied:
how could we avoid not adding anything that we know is past TTL? I think that's a better solution 👍

isaacs commented on src/security-archive/src/index.ts:
```diff
@@ -0,0 +1,297 @@
+   * Store current in-memory items to the database.
+   */
+  #storeItemsToDatabase(db: DatabaseSync): void {
+    const entries = this.dump()
```
This can be quite a lot of stuff, most of which hasn't changed.

It'd probably be better to gather up a batch of changed items, and only write back to the DB periodically, as items are updated.

ruyadorno replied:
right, we could tweak the implementation to hold a set of entries that were actually fetched from the API so that only those need to be updated.

ruyadorno replied:
done ✅

isaacs reviewed (dismissed):
This is a pretty good first pass, but I'd probably want to think through the data flows a bit more carefully to minimize massive load/save cycles, while still taking advantage of being able to batch up all the requests for a set of nodes in a graph. Especially, `LRUCache.load()` and `LRUCache.dump()` are pretty inefficient, and can get slow when the data grows.

I'd approach this as sort of a 3-level system, where you have the in-memory LRU, backed up by the database, backed further by the Socket API.

So, at initial load time, you give it a set of node IDs, and it tries to load those from the database. Any that it doesn't find, or where the TTL is expired, it fetches from the API server (perhaps in the background, and then saves back to the DB when it returns), so there's effectively zero delay, and never a full cache load/dump cycle.

ruyadorno commented:
> I'd approach this as sort of a 3-level system, where you have the in-memory LRU, backed up by the database, backed further by the Socket API.
> 
> So, at initial load time, you give it a set of node IDs, and it tries to load those from the database. Any that it doesn't find, or where the TTL is expired, it fetches from the API server (perhaps in the background, and then saves back to the DB when it returns), so there's effectively zero delay, and never a full cache load/dump cycle.

You just described the current implementation, so I'll take that as a nod of approval 😄 I think the only point of contention from your initial review is the idea of maybe leveraging `LRUCache.fetch` to have an ondemand load experience and I tried to explain why I avoided that in [this comment here](https://github.com/vltpkg/vltpkg/pull/518#discussion_r1996197735).

ruyadorno commented:
Added since the first pass:

- remodelled the db, separating `ttl` and `start` values as their own db column
- when selecting from the db, filter out expired entries
- validation that all packages required as part of the graph have their reports ready (`archive.ok` property)
- added a basic retry logic using `p-retry`
